Coffee Addicts project, v1.0

mihaigpm opened the pull request.

mihaigpm committed:
Project implementation

mihaigpm committed:
Naming fine tuning :)

mihaigpm committed:
Small refinements and tests (wip)

mihaigpm committed:
Validate location coordinates properly

mihaigpm committed:
Create library instead of executable since SPM executables are not testable.

mihaigpm committed:
Testing improvements

mihaigpm committed:
Readme fine tuning

mihaigpm committed:
Readme update

mihaigpm committed:
Update README.md

mihaigpm committed:
Update README.md

vladstanescu94 commented:
Hi, thanks for your PR, app looks great. I tried to run all the tests and one it's failing. Please check it out.

RalucaMihaela commented:
Hello. The solution works as expected. Would've been nice to have a PR description with instructions how to run app / tests.

vladstanescu94 commented on Sources/CoffeeAddictsLibrary/Commands/CoffeeShopsFinder.swift:
```diff
@@ -0,0 +1,62 @@
+        public func run() throws {
+            CoffeeShopsCSVDownloader.downloadCSVFileFromURLString(csvUrl) { csvString, status in
+                if status == .CSVFileDownloadSuccess {
+                    if let string = csvString {
```
Having multiple nested if statements looks weird. It's there a way to try to break the logic and make the code more readable?

mihaigpm replied:
Done, used a guard and moved part of the logic into a separate method.

vladstanescu94 commented on Sources/CoffeeAddictsLibrary/Commands/CoffeeShopsFinder.swift:
```diff
@@ -0,0 +1,62 @@
+                            let userLocation = CLLocation(latitude: xCoord, longitude: yCoord)
+                            let coffeeShops = CoffeeShopsLocationHelper.closestLocation(locations: result, closestToLocation: userLocation)
+                            
+                            if coffeeShops.count >= 3 {
```
Maybe it's better to extract this number into a variable that can tell what it's doing and not use magic numbers.

mihaigpm replied:
Done

RalucaMihaela commented on Sources/CoffeeAddictsLibrary/Helpers/CoffeeShopsCSVDownloader.swift:
```diff
@@ -0,0 +1,64 @@
+    static func downloadCSVFileFromURLString(_ urlString: String, completion: @escaping CSVResult) {
+        let sema = DispatchSemaphore(value: 0)
+
+        let documentsUrl:URL =  (FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first as URL?)!
```
It's not best to assume that we always have a file available by using force unwrap operator. Maybe can you use a `guard` to verify if there is a file?

mihaigpm replied:
Good catch, done.

RalucaMihaela commented on Sources/CoffeeAddictsLibrary/Helpers/CoffeeShopsCSVDownloader.swift:
```diff
@@ -0,0 +1,64 @@
+        let documentsUrl:URL =  (FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first as URL?)!
+        let destinationFileUrl = documentsUrl.appendingPathComponent(Constants.localCSVFileName)
+        
+        do {
```
I think this method has too much logic in it. Can we break it in smaller functions?

mihaigpm replied:
Done

RalucaMihaela commented on Sources/CoffeeAddictsLibrary/Helpers/CoffeeShopsCSVDownloader.swift:
```diff
@@ -0,0 +1,64 @@
+                    completion(nil, .CSVFileDownloadError)
+                }
+                
+                if let dir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first {
```
If I understand correctly, here you have the same url as you get on `documentsUrl` ? Can't you use that one?

mihaigpm replied:
Done

mihaigpm committed:
Code review changes

vladstanescu94 commented on Sources/CoffeeAddictsLibrary/Helpers/CoffeeShopsCSVParser.swift:
```diff
@@ -0,0 +1,48 @@
+        
+        let parsedCSV: [[String]] = string.components(separatedBy: "\n").map{ $0.components(separatedBy: ",") }
+        
+        if parsedCSV.count == 0 {
```
Can we use a guard clause here?

mihaigpm replied:
Done

RalucaMihaela commented on Sources/CoffeeAddictsLibrary/Helpers/CoffeeShopsCSVParser.swift:
```diff
@@ -0,0 +1,48 @@
+                continue
+            }
+            if let lat = Double(line[1].trimmingCharacters(in: .newlines)), let lng = Double(line[2].trimmingCharacters(in: .newlines)) {
+                if CoffeeShopsLocationHelper.locationIsValid(location: CLLocation(latitude: lat, longitude: lng)) {
```
Can you break also this one in multiple methods?

mihaigpm replied:
Partially done, I moved the lat/lng check into a separate method. I'd rather leave a for loop with several continue statements as it is, I find it more easily to follow where it will skip.

RalucaMihaela commented on Sources/CoffeeAddictsLibrary/Helpers/CoffeeShopsLocationHelper.swift:
```diff
@@ -0,0 +1,25 @@
+    }
+    
+    static func locationIsValid(location: CLLocation) -> Bool {
+        return (location.coordinate.latitude > -90 && location.coordinate.latitude < 90) && (location.coordinate.longitude > -180 && location.coordinate.longitude < 180)
```
I think is hard to read this return. Can you break it in multiple variables?

mihaigpm replied:
I left this as it is, since this serves a single purpose, coordinate validation, I personally think it's easier to follow the -90..90, -180..180 constraints this way

mihaigpm commented:
> Hi, thanks for your PR, app looks great. I tried to run all the tests and one it's failing. Please check it out.

Hi, thanks for the review. The boilerplate default test example was failing, I forgot to remove it. Removed it in commit cf74b43.

vladstanescu94 commented on Tests/CoffeeAddictsTests/CoffeeAddictsTests.swift:
```diff
@@ -0,0 +1,113 @@
+    }
+    
+    func testInvalidCSVFile() {
+//        let bundle = Bundle(for: type(of: self))
```
Did you forget to remove the comments?

mihaigpm replied:
Sort of, left them on purpose in case i find a good solution instead to read a file from bundle in an executable Swift Package. Currently the approach is buggy for Swift Packages that don't have a project file. 

I removed them now.

mihaigpm commented:
> Hello. The solution works as expected. Would've been nice to have a PR description with instructions how to run app / tests.

Hello, thanks for the review! I've added the running instructions in the README.md file inside this PR here: https://github.com/mihaigpm/coffee-addicts/pull/1/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5

I've updated it now to include testing instructions (cf74b43).

vladstanescu94 reviewed (changes requested):
Looks good, just some small changes 😄

RalucaMihaela reviewed (changes requested):
Great solution 👍, I left some comments, please look over them.

mihaigpm committed:
Code review changes.

alexandru-calinoiu commented on Sources/CoffeeAddictsLibrary/Utils/ParsingStatus.swift:
```diff
@@ -0,0 +1,18 @@
+
+import Foundation
+
+enum ParsingStatus: String {
```
There is a read flag the pops up as soon as I see folders like `utils` or `helpers` or `managers` ...

in your case I think `Constants` is fine (check out my general comment) but this enum might be better in another place, somewhere closer to the parser.

mihaigpm replied:
> There is a read flag the pops up as soon as I see folders like `utils` or `helpers` or `managers` ...
> 
> in your case I think `Constants` is fine (check out my general comment) but this enum might be better in another place, somewhere closer to the parser.

Besides my general comment, I agree that this could reside in the CSVParser itself or the the file could go in a separate folder together with the parser :)

I used to work on a project with two dozen enums and it was pretty difficult to debug and navigate through them because they were in the same files with other logic. 

So it’s just a reminiscence of that, I don’t hold to it or against it.

alexandru-calinoiu commented on Sources/CoffeeAddictsLibrary/Helpers/CoffeeShopsLocationHelper.swift:
```diff
@@ -0,0 +1,25 @@
+import Foundation
+import CoreLocation
+
+class CoffeeShopsLocationHelper {
```
Ha a `helper` class (see my comment relate to the utils folder), I will argue that this classes don't need to exist, for example this class does at least 2 things, some stuff relate do distance and validation.

One can pull a specialized class that does validation easily and even build a nice dsl for it.

What do you think? Are there good use cases for 'helper' classes?

mihaigpm replied:
> Ha a `helper` class (see my comment relate to the utils folder), I will argue that this classes don't need to exist, for example this class does at least 2 things, some stuff relate do distance and validation.
> 
> One can pull a specialized class that does validation easily and even build a nice dsl for it.
> 
> What do you think? Are there good use cases for 'helper' classes?

Replied to this in the general comment

alexandru-calinoiu commented on Sources/CoffeeAddictsLibrary/Helpers/CoffeeShopsCSVParser.swift:
```diff
@@ -0,0 +1,58 @@
+        var coffeeShops: [CoffeeShop] = []
+        
+        for line in parsedCSV {
+            if line.count != 3 {
```
I have a very particular linter that screams 'magic number' here, do you know about this code smell? And what we can do to avoid it?

mihaigpm replied:
> I have a very particular linter that screams 'magic number' here, do you know about this code smell? And what we can do to avoid it?

Totally agree, I’d move it into a ‘let expectedNumberOfLineParameters = 3’ variable.

Note to self: Add a linter from the beginning :) 

Which linter are you using, by the way?

alexandru-calinoiu replied:
Not sure what the options are in swift, I don't actively develop there :) but surely there must be one.

mihaigpm replied:
Got it, thx, I know a few ones, I thought you used one in particular

alexandru-calinoiu commented on Tests/CoffeeAddictsTests/CoffeeAddictsTests.swift:
```diff
@@ -0,0 +1,67 @@
+import class Foundation.Bundle
+@testable import CoffeeAddictsLibrary
+
+final class CoffeeAddictsTests: XCTestCase {
```
There a bunch of relevant test, I love the fact that you don't have individual test classes for each of the files, this made me think that you did some sort of tdd.

But you do test implementation details like helpers and validation, that is fine, on other way to do it was to use the actual entry point of your app (in this case the command). One benefit would have been that you could then vary the implementation and relay on the tests to make sure nothing is broken.

What is your experience with tests? When do you write tests and what type of tests do you usually write?

mihaigpm replied:
> There a bunch of relevant test, I love the fact that you don't have individual test classes for each of the files, this made me think that you did some sort of tdd.
> 
> But you do test implementation details like helpers and validation, that is fine, on other way to do it was to use the actual entry point of your app (in this case the command). One benefit would have been that you could then vary the implementation and relay on the tests to make sure nothing is broken.
> 
> What is your experience with tests? When do you write tests and what type of tests do you usually write?

In this particular case of a Swift Package (my first package written from scratch to be honest), I just found out while implementing, that you cannot unit test an executable package, only a library one. It seems that it’s a limitation due to the fact that by default a package doesn’t have a project file. If you add one then you break the cross-platform functionality. Therefore I had to pack most of the code into a library and then make the “main app” to use it. 

So due to this fact that I spent hours on figuring this out (not much documentation available for this atm), I made a trade-off between how much time I should/can spend on implementing this project and to show that writing tests is important. 

That being said, I did have in mind lots of other tests to write: 
- test valid/invalid command line arguments and match the expected error messages
- test all the possible CSV file scenarios by importing local sample files (empty file, malformed line endings, extra attributes, invalid attributes order etc), test distance calculations

I used the TDD approach on a new project at some point, but the process was pretty slow mainly due to constant requirements changing (not only I had to constantly rewrite code but also maintaining already existing tests). And since the business requirement was to get the product on the market fast, I gave up on TDD.

Nowadays, since in the past two years I worked mostly on projects that have to ship fast, I write unit tests for complex calculations, for dates manipulation and date formatting, files manipulation, session storage etc. I write some integration tests for regression purpose, testing a whole end-to-end flow.

alexandru-calinoiu reviewed (commented):
Hi Mihai, thanks for the submission.

The implementation indeed looks like it does it job, there is a decent separation of concerns going on.

Apart from my comments that is one other this that I will like to run by you. You seem to separate your classes by functionality, a mix between language specific functionality (like you will put extensions in a extensions folder) and design patterns (like you will put entities in a entities folder)

Do you know of other ways to do this organization? If one things about the domain how will you organize your code?

mihaigpm commented:
> Hi Mihai, thanks for the submission.
> 
> The implementation indeed looks like it does it job, there is a decent separation of concerns going on.
> 
> Apart from my comments that is one other this that I will like to run by you. You seem to separate your classes by functionality, a mix between language specific functionality (like you will put extensions in a extensions folder) and design patterns (like you will put entities in a entities folder)
> 
> Do you know of other ways to do this organization? If one things about the domain how will you organize your code?

Thank you for the detailed review and for bringing good topics for discussion.

First I want to say that this project could have been written in a few different ways and that there is definitely room for improvement!

Regarding the folder structure/code organization, normally I prefer to start organizing by features/functionalities, then inside each feature I’d create folders for domain specific models/entities, views (and/or view controllers), business logic and domain specific helpers and extensions. All these contained in a “App (or Features)” level/folder. Then on the same horizontal level I have a Networking layer if needed, a Navigation/Coordination layer and… an Utils/Extensions one which contains only generic app-wide helpers.

While I agree that a helpers/utils folder has a potential to act as a garbage can for everything, I think with a sense of responsibility from the developer’s side it’s useful for general reusable pieces of code that can be used over more features.

For the location class example that you gave in a comment below, it’s arguable (and I thought to) if I could have added the validation and distance calculation methods as part of a CLLocation extension. But regardless, its purpose is generic, for example you could reuse the lat/lng validation everywhere you’d need, it’s not domain specific. 

My personal opinion is that one should use a structure adapted to the project needs. Of course, without sacrificing readability or performance! For example, if you take over an existing large project, it’s less likely and might impact the velocity of team members if you start refactoring everything to the style you are used to. I read somewhere a nice thing, that you should try to leave a piece of code in a better state that how you find it. But gradually, not at once.

Note: I do have an argument against something like utils, in case of modularisation. Assume you have module A and module B and both use something from an Utils module. Then yes, you’d have Utils as a dependency for both A and B which means B might import stuff that are needed only for A and vice-versa. But this can be a whole topic to discuss for hours :)

mihaigpm committed:
Code review changes
